Lilu/Headers/kern_util.hpp

```diff
@@ -944,4 +945,279 @@ class evector_base {
+	 */
+	static void deleter(CircularBuffer<T> *buffer NONNULL) {
+		Buffer::deleter(buffer->storage);
+		buffer->deinit();
+		delete buffer;
+	}
+	
+	/**
+	 *  Destory the given circular buffer if it is non-null and set it to nullptr
+	 *
+	 *  @param buffer A nullable circular buffer returned by `CircularBuffer::withCapacity()`.
+	 *  @note This function mimics the macro `OSSafeReleaseNULL()`.
+	 */
+	static void safeDeleter(CircularBuffer<T> *&buffer) {
+		if (buffer != nullptr) {
+			deleter(buffer);
+			buffer = nullptr;
+		}
+	}
+	
+	/**
+	 *  Check whether the circular buffer is empty
+	 *
+	 *  @return `true` if the buffer is empty, `false` otherwise.
+	 */
+	bool isEmpty() {
+		IORecursiveLockLock(lock);
+		bool retVal = (count == 0) && (indexr == indexw);
+		IORecursiveLockUnlock(lock);
+		return retVal;
+	}
+	
+	/**
+	 *  Check whether the circular buffer is full
+	 *
+	 *  @return `true` if the buffer is full, `false` otherwise.
+	 */
+	bool isFull() {
+		IORecursiveLockLock(lock);
+		bool retVal = (count == size) && (indexr == indexw);
+		IORecursiveLockUnlock(lock);
+		return retVal;
+	}
+	
+	/**
+	 *  Get the number of elements in the circular buffer
+	 *
+	 *  @return The current number of elements in the buffer.
+	 */
+	IOItemCount getCount() {
+		IORecursiveLockLock(lock);
+		IOItemCount retVal = count;
+		IORecursiveLockUnlock(lock);
+		return retVal;
+	}
+	
+	/**
+	 *  Write the given element to the circular buffer
+	 *
+	 *  @param element The element to write
+	 *  @return `true` on success, `false` if the buffer is full.
+	 */
+	bool push(const T &element) {
+		IORecursiveLockLock(lock);
+		if (isFull()) {
+			IORecursiveLockUnlock(lock);
+			return false;
+		}
+		storage[indexw] = element;
+		indexw += 1;
+		indexw %= size;
+		count += 1;
+		IORecursiveLockUnlock(lock);
+		return true;
+	}
+	
+	/**
+	 *  Read the next element from the circular buffer
+	 *
+	 *  @param element The element read from the buffer
+	 *  @return `true` on success, `false` if the buffer is empty.
+	 */
+	bool pop(T& element) {
+		IORecursiveLockLock(lock);
+		if (isEmpty()) {
+			IORecursiveLockUnlock(lock);
+			return false;
+		}
+		element = storage[indexr];
+		indexr += 1;
+		indexr %= size;
+		count -= 1;
+		IORecursiveLockUnlock(lock);
+		return true;
+	}
+};
+
+/**
+ *  Wrap an object that is not an instance of OSObject
+ */
+class EXPORT OSObjectWrapper: public OSObject {
+	/**
+	 *  Constructors & Destructors
+	 */
+	OSDeclareDefaultStructors(OSObjectWrapper);
+	
+	using super = OSObject;
+	
+	/**
+	 *  Wrapped object
+	 */
+	void *object {nullptr};
+	
+public:
+	/**
+	 *  Initialize the wrapper with the given object
+	 *
+	 *  @param object The wrapped object that is not an `OSObject`
+	 *  @return `true` on success, `false` otherwise.
+	 */
+	EXPORT bool init(void *object);
+	
+	/**
+	 *  Reinterpret the wrapped object as the given type
+	 *
+	 *  @return The wrapped object of the given type.
+	 */
+	template <typename T>
+	T *get() {
```

Comment:
Now to think of it, will not making this class templated simplify its accessor?

Reply:
Unfortunately, you can't make the class itself templated, because `OSMetaClass` does not allow it. 
Subclasses of `OSObject` cannot be a template class nor inherit from multiple base classes.

Reply:
Right.